Story block: Post Images support

aforcier opened the pull request:
#### Changes proposed in this Pull Request:

Adds support for the Story block to `Jetpack_PostImages`, allowing the Open Graph image tags (`og:image` and `twitter:image`) to be set for posts containing the story block. This also enables themes that support it (e.g. Cubic) to have a preview image for story posts in the post gallery, without requiring a featured image.

~I'm also proposing a tweak to the open graph image selection logic, so it selects the first 'good' image in a post instead of the last (the Twitter card tag logic already does this).~ Extracted that part to #17598.

The image tags can be inspected by viewing the source of a post page, they look like this:

```html
<!-- Jetpack Open Graph Tags -->
<meta property="og:type" content="article" />
<meta property="og:title" content="A story post" />
<meta property="og:url" content="https://example.com/2020/10/20/a-story-post/" />
<meta property="og:description" content="Visit the post for more." />
<meta property="article:published_time" content="2020-10-20T06:39:37+00:00" />
<meta property="article:modified_time" content="2020-10-20T06:39:37+00:00" />
<meta property="og:site_name" content="alexforcier" />
<meta property="og:image" content="https://i2.wp.com/example.com/wp-content/uploads/2020/10/cropped-WP-Icon.png?fit=512%2C512&amp;ssl=1" />
<meta property="og:image:width" content="512" />
<meta property="og:image:height" content="512" />
<meta property="og:locale" content="en_US" />
<meta name="twitter:text:title" content="A story post" />
<meta name="twitter:image" content="https://i2.wp.com/example.com/wp-content/uploads/2020/10/cropped-WP-Icon.png?fit=240%2C240&amp;ssl=1" />
<meta name="twitter:card" content="summary" />
<meta name="twitter:description" content="Visit the post for more." />
```

TODO:
* [x] ~Currently incorrectly sets a VideoPress video as the social media image when that's the first slide of a story~
  * --> Fixed to extract the poster image for VideoPress videos
* [x] ~Fixing the above relies on using VideoPress poster images, and there's currently an issue where we either need to Photon-wrap them or make a server-side change - TBD~
   * --> This was patched server-side - I'll open a separate PR soon to propose some broader changes, but all that matters for this PR is that the `w=....` parameter automatically appended to `twitter:image` links no longer 404s for VideoPress poster images

I had to add special logic for supporting WordPress.com, for whatever reason the media metadata doesn't contain VideoPress data, but instead the `thumb` parameter is available. If I missed something here I'm very happy to correct it.

#### Does this pull request change what data or activity we track or use?

No.

#### Testing instructions:

On a Jetpack site, and on a WordPress.com simple site (the code path is different between the two):

Inspect the image tags for various kinds of stories and verify that the correct image is used:

(These apply to the `twitter:image` tag - the `og:image` tag will follow the same logic but use the _last_ slide with a valid image. #17598 will change the og tag to be in line with the twitter tag.)

* A story where the first slide is an image -> the image tag should have that image
* A story where the first slide is a VideoPress video -> the image tag should have that video's poster
* A story where the first slide is a non-VideoPress video -> the first slide that is an image (or a VideoPress video) should be used - otherwise no image if all slides are non-VideoPress videos

jeherve commented on class.jetpack-post-images.php:
```diff
@@ -806,6 +806,66 @@ public static function get_attachment_data( $attachment_id, $post_url = '', $wid
+			$url         = $meta['videopress']['poster'];
+			$meta_width  = $meta['videopress']['width'];
+			$meta_height = $meta['videopress']['height'];
+		} elseif ( ! empty( $meta['thumb'] ) ) {
```
Could you clarify with an inline comment what scenario this is meant to catch?

aforcier replied:
Done in b2011c2, let me know if you have anything better in mind for the comment 👍

jeherve reviewed (changes requested):
Do you think you could add tests for this in `tests/php/media/test-class.jetpack-post-images.php`?

> I'm also proposing a tweak to the open graph image selection logic, so it selects the first 'good' image in a post instead of the last (the Twitter card tag logic already does this).

The change makes sense, but it worries me a bit. I wonder why this was implemented that way (it seems it's been that way since the beginning), and I worry that I may be missing something here.

What would you say about moving this into its own PR, with a set of tests to go with it (see `tests/php/test_class.functions.opengraph.php`)?

Tug reviewed (approved):
Change works well 👍 , I didn't understand why it was not working at first, I had to enable sharing in jetpack settings!

About the change in `class.jetpack-post-images.php`  I wonder if we could not simply update `get_attachment_data`, not sure there is any benefit in having a special case for story.

@jeherve According to the [PR that introduced this](https://github.com/Automattic/jetpack/pull/2291) it looks like there was not particular reason for this change, it was done to update the format returned by `jetpack_og_get_image`

aforcier commented:
> Do you think you could add tests for this in tests/php/media/test-class.jetpack-post-images.php?

Added tests in dd6f5b5 ✅ I'll be more proactive about looking for existing test classes in the future 😀 

> The change makes sense, but it worries me a bit. I wonder why this was implemented that way (it seems it's been that way since the beginning), and I worry that I may be missing something here.
> What would you say about moving this into its own PR, with a set of tests to go with it (see tests/php/test_class.functions.opengraph.php)?

That sounds reasonable. I undid the change in efd2391 and moved that commit to its own PR here: https://github.com/Automattic/jetpack/pull/17598. I haven't written tests for that yet, I'll get to them before I un-draft the PR. 👍 

> About the change in class.jetpack-post-images.php I wonder if we could not simply update get_attachment_data, not sure there is any benefit in having a special case for story.

There wasn't a special case really except that Story seems to be the first block considered in that class that can have mixed media. I added a function just for Stories because I was trying to impact existing code as little as possible, but it probably makes more sense to consolidate.

I'm not sure what you think about that @jeherve  - I made the change in 97b94aa, we can revert it if needed.

aforcier commented on class.jetpack-post-images.php:
```diff
@@ -806,6 +806,66 @@ public static function get_attachment_data( $attachment_id, $post_url = '', $wid
+			return false;
+		}
+
+		if ( ! empty( $meta['videopress'] ) ) {
```
Done, see my top level comment here: https://github.com/Automattic/jetpack/pull/17198#issuecomment-715660755

jeherve reviewed (approved):
I wasn't able to test with a VideoPress video (I could not find a way to insert such a video from the mobile app), but other than that it looks good to me. 👍

aforcier commented:
> I wasn't able to test with a VideoPress video (I could not find a way to insert such a video from the mobile app), but other than that it looks good to me. 👍

Thanks @jeherve! FYI if VideoPress is enabled for a site any video uploaded from the mobile apps _should_ automatically be handled by VideoPress. I'll go ahead and merge this but if you're having a different experience let me know and that might be a separate issue we might want to look into.

jeherve commented:
> if VideoPress is enabled for a site any video uploaded from the mobile apps should automatically be handled by VideoPress

That didn't quite work like that in my test; the video was uploaded to my local (Jetpack) site. I'm happy to open an issue elsewhere about this if needed.

aforcier commented:
> That didn't quite work like that in my test; the video was uploaded to my local (Jetpack) site. I'm happy to open an issue elsewhere about this if needed.

Interesting! I'll DM you for some more information.

aforcier commented:
r215877-wpcom
